artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerContainer_ContainerManagementInterface-w1/source/src/main/java/org/opentosca/artifacttemplates/dockercontainer/DockerContainer.java

```diff
@@ -0,0 +1,178 @@
+                LOG.info("Container is available");
+                return;
+            } catch (InterruptedException e) {
+                LOG.error("Could not await availability", e);
+                error = e;
+                try {
+                    Thread.sleep(1000);
+                } catch (InterruptedException e1) {
+                    LOG.error("Could not await availability", e1);
+                }
+            }
+        }
+        if (error != null) {
+            throw error;
+        }
+    }
+
+    public String execCommand(String command) throws InterruptedException {
+        LOG.info("Executing shell command: {}", command);
+
+        try (OutputStream stream = getStringOutputStream()) {
+            ExecCreateCmdResponse execCreateCmdResponse = client.execCreateCmd(containerId)
+                    .withTty(true)
+                    .withAttachStdout(true)
+                    .withCmd("/bin/bash", "-c", command)
+                    .exec();
+
+            client.execStartCmd(execCreateCmdResponse.getId())
+                    .withTty(true)
+                    .withDetach(false)
+                    .exec(new ExecStartResultCallback(stream, System.err))
+                    .awaitCompletion();
+
+            String result = stream.toString();
+            LOG.info("Execution log:\n {}", result);
+            return result;
+        } catch (IOException e) {
+            LOG.error("Could not execute command", e);
+        }
+
+        return "";
+    }
+
+    // TODO: replace this with other new function
+    public String replaceHome(String commandString, boolean all) {
+        if (commandString.contains("~")) {
+            try {
+                String pwd = execCommand("pwd").trim();
+                System.out.println("Replaced ~ with user home ('" + pwd + "'): '" + commandString + "'");
+                return all ? commandString.replaceAll("~", pwd) : commandString.replaceFirst("~", pwd);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+        }
+        return commandString;
+    }
+
+    public String replaceHome(String command) throws InterruptedException {
+        if (command.contains("~/")) {
+            String pwd = execCommand("pwd").trim();
+            String replaced = command.replaceAll("~/", pwd);
+            LOG.info("Replaced '~' in '{}' with home '{}' which results in '{}'", command, pwd, replaced);
+            return replaced;
+        }
+        return command;
+    }
+
+    public void uploadFile(String source, String target) throws InterruptedException {
+        LOG.info("Uploading file '{}' to '{}'...", source, target);
+
+        String folders = target.substring(0, target.lastIndexOf('/'));
+        LOG.info("Creating remote directory '{}'", folders);
+        execCommand("mkdir -p " + folders);
+
+        LOG.info("Copy file to container");
+        client.copyArchiveToContainerCmd(containerId)
```

Comment:
At this point the execution stops. 
There seems to be a problem with coyping the source to the target.